src/cargo/core/resolver/mod.rs

```diff
@@ -247,158 +193,245 @@ pub fn resolve(summary: &Summary, method: &Method,
-                     mut deps: slice::Iter<'a, DepInfo>,
-                     cur: usize,
-                     finished: &mut FnMut(Box<Context>, &mut Registry) -> ResolveResult)
-                     -> ResolveResult {
-    let &(dep, ref candidates, ref features) = match deps.next() {
-        Some(info) => info,
-        None => return finished(cx, registry),
-    };
+/// dependency graph, cx.resolve is returned.
+fn activate_deps_loop(mut cx: Context,
+                      registry: &mut Registry,
+                      top: Rc<Summary>,
+                      top_method: &Method) -> CargoResult<Resolve> {
+    let mut backtrack_stack = Vec::new();
+    let mut remaining_deps = Vec::new();
+    remaining_deps.extend(try!(activate(&mut cx, registry, top, &top_method)));
+    loop {
+        // Retrieves the next dependency to try, from `remaining_deps`.
+        let frame = match remaining_deps.pop() {
+            None => break,
+            Some(mut deps_frame) => {
+                match deps_frame.remaining_siblings.next() {
+                    None => {
+                        cx.visited.remove(&deps_frame.id);
+                        continue
+                    }
+                    Some((cur, (dep, candidates, features))) => {
+                        let parent = deps_frame.parent.clone();
+                        remaining_deps.push(deps_frame);
+                        (parent, cur, dep, candidates, features)
+                    }
+                }
+            }
+        };
+        let (mut parent, mut cur, mut dep, candidates, features) = frame;
+        assert!(!remaining_deps.is_empty());
 
-    let method = Method::Required {
-        dev_deps: false,
-        features: features,
-        uses_default_features: dep.uses_default_features(),
-    };
+        let method = Method::Required {
+            dev_deps: false,
+            features: &features,
+            uses_default_features: dep.uses_default_features(),
+        };
 
-    let prev_active = cx.prev_active(dep);
-    trace!("{}[{}]>{} {} candidates", parent.name(), cur, dep.name(),
-           candidates.len());
-    trace!("{}[{}]>{} {} prev activations", parent.name(), cur,
-           dep.name(), prev_active.len());
-
-    // Filter the set of candidates based on the previously activated
-    // versions for this dependency. We can actually use a version if it
-    // precisely matches an activated version or if it is otherwise
-    // incompatible with all other activated versions. Note that we define
-    // "compatible" here in terms of the semver sense where if the left-most
-    // nonzero digit is the same they're considered compatible.
-    let my_candidates = candidates.iter().filter(|&b| {
-        prev_active.iter().any(|a| a == b) ||
-            prev_active.iter().all(|a| {
-                !compatible(a.version(), b.version())
-            })
-    });
+        let prev_active = cx.prev_active(&dep).to_vec();
```

Comment:
I think you can get away with an extra scope here which kills the borrow right after `my_candidates` is assembled.